network/stub/network.go

```diff
@@ -268,7 +279,13 @@ func (n *Network) sendToAllTargets(m *PendingMessage, syncOnProcess bool) error
 		receiverNetwork.seen(key)
 
 		// finds the engine of the targeted Network
-		receiverEngine, ok := receiverNetwork.engines[m.Channel]
+		var ok bool
+		var receiverEngine network.MessageProcessor
+		if receiverNetwork == n {
+			receiverEngine, ok = n.getEngineLocked(m.Channel)
+		} else {
+			receiverEngine, ok = receiverNetwork.GetEngine(m.Channel)
```

Comment:
This is a thread safe data race ....

1.  GetEngine returns a valid engine
2. context switch & the engine got unregister
3. context switch back to current goroutine and process

Reply:
you probably want to replace GetEngine with ProcessWithEngine (The lock should be held while Process() is using the engine).

Reply:
done!  I'm a bit concerned about deadlocks given that we now hold multiple locks but it tis no worse than it was.